renderdoc/driver/vulkan/vk_counters.cpp

```diff
@@ -55,43 +78,142 @@ void VulkanReplay::DescribeCounter(uint32_t counterID, CounterDescription &desc)
+      desc.description = "Number of samples that passed depth/stencil test.";
+      desc.resultByteWidth = 8;
+      desc.resultCompType = eCompType_UInt;
+      desc.units = eUnits_Absolute;
+      break;
+    case eCounter_VSInvocations:
+      desc.name = "VS Invocations";
+      desc.description = "Number of times a vertex shader was invoked.";
+      desc.resultByteWidth = 8;
+      desc.resultCompType = eCompType_UInt;
+      desc.units = eUnits_Absolute;
+      break;
+    case eCounter_GSInvocations:
+      desc.name = "GS Invocations";
+      desc.description = "Number of times a geometry shader was invoked.";
+      desc.resultByteWidth = 8;
+      desc.resultCompType = eCompType_UInt;
+      desc.units = eUnits_Absolute;
+      break;
+    case eCounter_TCSInvocations:
+      desc.name = "TCS Invocations";
+      desc.description = "Number of times a tesselation control shader was invoked.";
+      desc.resultByteWidth = 8;
+      desc.resultCompType = eCompType_UInt;
+      desc.units = eUnits_Absolute;
+      break;
+    case eCounter_TESInvocations:
+      desc.name = "TES Invocations";
+      desc.description = "Number of times a tesselation evaluation shader was invoked.";
+      desc.resultByteWidth = 8;
+      desc.resultCompType = eCompType_UInt;
+      desc.units = eUnits_Absolute;
+      break;
+    case eCounter_PSInvocations:
+      desc.name = "PS Invocations";
+      desc.description = "Number of times a pixel shader was invoked.";
+      desc.resultByteWidth = 8;
+      desc.resultCompType = eCompType_UInt;
+      desc.units = eUnits_Absolute;
+      break;
+    case eCounter_CSInvocations:
+      desc.name = "CS Invocations";
+      desc.description = "Number of times a compute shader was invoked.";
+      desc.resultByteWidth = 8;
+      desc.resultCompType = eCompType_UInt;
+      desc.units = eUnits_Absolute;
+      break;
+    default:
+      desc.name = "Unknown";
+      desc.description = "Unknown counter ID";
+      desc.resultByteWidth = 0;
+      desc.resultCompType = eCompType_None;
+      desc.units = eUnits_Absolute;
+      break;
   }
 }
+
 struct VulkanGPUTimerCallback : public VulkanDrawcallCallback
 {
-  VulkanGPUTimerCallback(WrappedVulkan *vk, VulkanReplay *rp, VkQueryPool qp)
-      : m_pDriver(vk), m_pReplay(rp), m_QueryPool(qp)
+  VulkanGPUTimerCallback(WrappedVulkan *vk, VulkanReplay *rp, VkQueryPool tsqp, VkQueryPool occqp,
+                         VkQueryPool psqp)
+      : m_pDriver(vk),
+        m_pReplay(rp),
+        m_TimeStampQueryPool(tsqp),
+        m_OcclusionQueryPool(occqp),
+        m_PipeStatsQueryPool(psqp)
   {
     m_pDriver->SetDrawcallCB(this);
   }
   ~VulkanGPUTimerCallback() { m_pDriver->SetDrawcallCB(NULL); }
   void PreDraw(uint32_t eid, VkCommandBuffer cmd)
   {
-    ObjDisp(cmd)->CmdWriteTimestamp(Unwrap(cmd), VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, m_QueryPool,
-                                    (uint32_t)(m_Results.size() * 2 + 0));
+    if(m_OcclusionQueryPool != NULL)
```

Comment:
Vulkan handles should be compared against `VK_NULL_HANDLE`, not `NULL`, since they're not necessarily pointers.